server/server.go

```
@@ -0,0 +1,146 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"log"
+	"net/http"
+	"sync"
+)
+
+// Errors : to display errors from the server
+type Errors struct {
+	Errs []string
+}
+
+
+// Info : data store
+type Info struct {
+	sync.Mutex
+	Data map[string]string
+}
+
+var (
+	httpAddr = flag.String("addr", ":8098", "server addr")
+)
+
+func main() {
+	flag.Parse()
+
+	RunServer()
+}
+
+//New : create data data store
+func New() *Info {
+	return &Info{
+		Data: make(map[string]string),
+	}
+}
+
+// RunServer : start server
+func RunServer() {
+	http.HandleFunc("/get", handleGet)
+	http.HandleFunc("/set", handleSet)
+
+	log.Println("Start server at :8098 ...")
+	err := http.ListenAndServe(*httpAddr, nil)
+	if err != nil {
+		log.Fatal("Error happend", err.Error())
+	}
+}
+
+var data = New()
+
+func (inf *Info) set(phone, name string) {
+	inf.Lock()
+	inf.Data[phone] = name
+	inf.Unlock()
+}
+
+func (inf *Info) get(phone string) (string, bool) {
+	inf.Lock()
+	defer inf.Unlock()
+	val, ok := inf.Data[phone]
+	return val, ok
+}
+
+func handleSet(w http.ResponseWriter, r *http.Request) {
+	var name, phone string
+	if name, phone = r.FormValue("name"), r.FormValue("phone"); name == "" || phone == "" {
+		errs := Errors{}
+
+		if phone == "" {
+			errs.Errs = append(errs.Errs, "undefined param phone")
+		}
+		if name == "" {
+			errs.Errs = append(errs.Errs, "undefined param name")
+		}
+
+		w.WriteHeader(http.StatusBadRequest)
+		resp, err := json.Marshal(errs)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		w.Write(resp)
+		return
+	}
+
+	data.set(phone, name)
+
+	val, _ := data.get(phone)
+	response, err := json.Marshal(val)
+	if err != nil {
+		log.Println("http.Error")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+	w.Write(response)
```

I assume here is error ignored
Please never ignore errors